File: crates/testrunner/harness/php/check.php
<?php
// Vybe test harness — PHP.
//
// PHP does not pair like Go or JavaScript. `echo` is a statement, not a call,
// and it appends no newline: of 15,477 echos in the corpus only 1,248 carry a
// "\n", so consecutive echos land on ONE line and there is no i-th print to
// match to an i-th expected line.
//
// So instead of pairing, the whole program runs inside an output buffer and its
// entire output is compared once. That is exact rather than heuristic, and it
// needs no analysis of the program at all.
//
// The verdict is the EXIT CODE; the diagnostic is printed before throwing,
// because an uncaught error renders as `[object]` and says nothing.

function __vybe_check($got, $want) {
    // Match the Rust harness's normalisation: strip \r, then drop trailing
    // newlines (it split on "\n" and popped empty trailing elements).
    $got = str_replace("\r", "", $got);
    $got = rtrim($got, "\n");
    if ($got !== $want) {
        echo "FAIL: want [" . $want . "] got [" . $got . "]\n";
        throw new Exception("assertion failed");
    }
    // Replay the program's own output so running the file by hand still
    // behaves like the program it was extracted from.
    echo $got;
    if ($got !== "") {
        echo "\n";
    }
}

File: crates/testrunner/harness/lua/check.lua
-- Vybe test harness — Lua.
--
-- Real Lua: this runs under `lua` unchanged. Like the C, COBOL and Fortran
-- harnesses it documents the shape the emitter produces rather than providing
-- a function to splice, because the check has to be at the CALL SITE.
--
-- WHY NOT COLLECT THE OUTPUT.
-- Every other collecting harness (ruby, java, dart, pascal, php) intercepts
-- the language's print. Lua cannot, under Vybe — measured, all three forms
-- fall through to the real stdout:
--
--   function print(...) ... end     -- builtin still used
--   print = function(...) ... end   -- builtin still used
--   _G.print = function(...) ... end -- builtin still used
--
-- (`io.write` is also undefined.) So each `print` is compared where it stands.
--
-- WHY ONLY THE FIRST PRINT.
-- The corpus asserts through `run_lua_one`, which takes `run_lua(src)` and
-- returns only its FIRST line. A program that prints more is not asserting
-- those lines, so checking them would invent expectations the test never had.
-- The counter still advances on every print, which is what lets a print inside
-- a LOOP be handled: the check fires on print #1 wherever it occurs.
--
-- The emitted shape, for `print(a .. "," .. b)` expecting "2,1":
--
--   local __w1 = "2,1"
--   local __i = 0
--   ...
--   do local __t = tostring(a .. "," .. b); __i = __i + 1
--      if __i == 1 and __t ~= __w1 then
--        error("FAIL: want [" .. __w1 .. "] got [" .. __t .. "]") end end
--   ...
--   if __i == 0 then error("FAIL: no output") end
--
-- Multiple arguments join with a TAB, which is what Lua's `print` does —
-- verified identical in both runtimes.
--
-- Failure is `error(...)`, which exits non-zero in both (lua 1, Vybe 1) and
-- carries its message, so the expected and actual values survive.

local __w1 = "2,1"
local __i = 0

local a, b = 1, 2
a, b = b, a

do
  local __t = tostring(a .. "," .. b)
  __i = __i + 1
  if __i == 1 and __t ~= __w1 then
    error("FAIL: want [" .. __w1 .. "] got [" .. __t .. "]")
  end
end

if __i == 0 then error("FAIL: no output") end
